### PR TITLE
Remove zone history check

### LINKED ISSUE #362: Remove calls to /zones/id/history from the python tests
**Describe the bug**
vinyldns_python has a function `get_zone_history` calling /zones/id/history. That endpoint doesnt exist in the api.

We are calling the above in `wait_until_zone_change_status`, so any calls to that function are waiting the full time, not until the status has changed.

In short, we need to remove get_zone_history and update or remove anything calling it.

### PR DESCRIPTION
**Pull request by britneywright**
Fixes #362

Changes in this pull request:
- removed `get_zone_history method` and `get_zone_change` methods
- update `wait_until_zone_change_status` method to check the status of the most recent zone change

### CODE REVIEW

File: modules/api/functional_test/vinyldns_python.py [hunk 1]
```diff
@@ -704,22 +671,20 @@ def wait_until_zone_change_status(self, zone_change, expected_status):
         """
         Waits until the zone change status matches the expected status
```
Comment (rebstar6) [inline modules/api/functional_test/vinyldns_python.py]: looking at usage, we are only ever calling this with status='Synced'. Because of that, i would probably change this to `wait_until_zone_change_synced`, meaning you can cut out early if the status is `Failed`

File: modules/api/functional_test/vinyldns_python.py [hunk 2]
```diff
@@ -704,22 +671,20 @@ def wait_until_zone_change_status(self, zone_change, expected_status):
...
+        def change_id_match(change):
+            return change[u'id'] == zone_change[u'id']
 
+        while change[u'status'] != expected_status and retries > 0:
+            latest_changes = self.list_zone_changes(change['zone']['id'])
+            if type(latest_changes).__name__  == 'dict' and latest_changes[u'zoneChanges']:
+                matching_changes = filter(change_id_match, latest_changes[u'zoneChanges'])
                 if len(matching_changes) > 0:
                     change = matching_changes[0]
+            else:
```
Comment (rebstar6) [inline modules/api/functional_test/vinyldns_python.py]: shouldnt need this else, its not doing anything

File: modules/api/functional_test/vinyldns_python.py [hunk 3]
```diff
@@ -704,22 +671,20 @@ def wait_until_zone_change_status(self, zone_change, expected_status):
...
 
-            if u'zoneChanges' in history:
-                zone_changes = history[u'zoneChanges']
-                matching_changes = filter(change_id_match, zone_changes)
+        def change_id_match(change):
+            return change[u'id'] == zone_change[u'id']
 
+        while change[u'status'] != expected_status and retries > 0:
+            latest_changes = self.list_zone_changes(change['zone']['id'])
+            if type(latest_changes).__name__  == 'dict' and latest_changes[u'zoneChanges']:
```
Comment (rebstar6) [inline modules/api/functional_test/vinyldns_python.py]: do we need this type check? id think ` if u'zoneChanges' in latest_changes:` would be ok

File: modules/api/functional_test/vinyldns_python.py [hunk 4]
```diff
@@ -704,22 +671,20 @@ def wait_until_zone_change_status(self, zone_change, expected_status):
...
-
         change = zone_change
         retries = MAX_RETRIES
-        while change[u'status'] != expected_status and retries > 0:
-            history = self.get_zone_history(zone_change[u'zone'][u'id'])
 
-            if u'zoneChanges' in history:
-                zone_changes = history[u'zoneChanges']
-                matching_changes = filter(change_id_match, zone_changes)
+        def change_id_match(change):
```
Comment (mitruly) [inline modules/api/functional_test/vinyldns_python.py]: This isn't explicitly needed and `change` here shadows the variable in this function. Piggybacking off of what @rebstar6 said, this is a bit cleaner:

```
    def wait_until_zone_change_status(self, zone_change, expected_status):
        """
        Waits until the zone change status matches the expected status
        """
        change = zone_change
        retries = MAX_RETRIES

        while change[u'status'] != expected_status and retries > 0:
            latest_changes = self.list_zone_changes(change['zone']['id'])
            if u'zoneChanges' in latest_changes:
                matching_changes = filter(lambda c: c[u'id'] == zone_change[u'id'], latest_changes[u'zoneChanges'])
                if len(matching_changes) > 0:
                    change = matching_changes[0]
            time.sleep(RETRY_WAIT)
            retries -= 1

        return change[u'status'] == expected_status
```

File: modules/api/functional_test/vinyldns_python.py (hunk 3 above)
Comment (mitruly) [reply modules/api/functional_test/vinyldns_python.py]: +1 here. The scripting aspect of Python allows us to simply check if a value is defined before using it rather than checking a type.

File: modules/api/functional_test/vinyldns_python.py [hunk 5]
```diff
@@ -700,30 +667,23 @@ def wait_until_recordset_deleted(self, zone_id, record_set_id, **kwargs):
...
+        latest_change = zone_change
         retries = MAX_RETRIES
-        while change[u'status'] != expected_status and retries > 0:
-            history = self.get_zone_history(zone_change[u'zone'][u'id'])
-
-            if u'zoneChanges' in history:
-                zone_changes = history[u'zoneChanges']
-                matching_changes = filter(change_id_match, zone_changes)
 
+        while latest_change[u'status'] != 'Synced' and retries > 0:
```
Comment (rebstar6) [inline modules/api/functional_test/vinyldns_python.py]: ^^ OR `latest_change[u'status'] != 'Failed'`. if its failed we can cut out

File: modules/api/functional_test/vinyldns_python.py [hunk 6]
```diff
@@ -700,30 +667,23 @@ def wait_until_recordset_deleted(self, zone_id, record_set_id, **kwargs):
...
+            if u'zoneChanges' in changes:
+                matching_changes = filter(lambda change: change[u'id'] == zone_change[u'id'], changes[u'zoneChanges'])
                 if len(matching_changes) > 0:
-                    change = matching_changes[0]
+                    latest_change = matching_changes[0]
             time.sleep(RETRY_WAIT)
             retries -= 1
 
-        return change[u'status'] == expected_status
+        return latest_change[u'status'] == 'Synced'
```
Comment (rebstar6) [inline modules/api/functional_test/vinyldns_python.py]: i kinda think we should `assert(latest_change[u'status'] == 'Synced')` here (or whatever that syntax is). weirdly, we call this wait_for but it does not fail if the condition is not met, just returns false

**Comment by britneywright** [comment]
Looks like the func tests are now taking 15 minutes to complete in Travis https://travis-ci.org/vinyldns/vinyldns/builds/463580219. It was down to 11 minutes before the most recent commits. For added reference Master takes about 13 minutes to run the func tests now.

File: modules/api/functional_test/vinyldns_python.py (hunk 6 above)
Comment (britneywright) [reply modules/api/functional_test/vinyldns_python.py]: for some reason adding the assertion makes `test_sync_zone_success` fail in Travis (https://travis-ci.org/vinyldns/vinyldns/jobs/463879028#L9042). It never fails locally though. But it didn't fail when I also had the `latest_change[u'status'] != 'Failed'` in the while statement of the function though, maybe it's a timing issue?

File: modules/api/functional_test/live_tests/zones/create_zone_test.py [hunk 7]
```diff
@@ -63,7 +63,9 @@ def test_create_zone_success(shared_zone_test_context):
         }
         result = client.create_zone(zone, status=202)
         result_zone = result['zone']
-        client.wait_until_zone_change_status(result, 'Synced')
+        client.wait_until_zone_change_status_synced(result)
+
+        time.sleep(1)
```
Comment (rebstar6) [inline modules/api/functional_test/live_tests/zones/create_zone_test.py]: why are we sleeping?

Comment (britneywright) [reply modules/api/functional_test/live_tests/zones/create_zone_test.py]: this was one of the tests that kept randomly failing in Travis. I thought it might be a timing issue.

File: modules/api/functional_test/vinyldns_python.py [hunk 8]
```diff
@@ -764,6 +724,23 @@ def wait_until_zone_exists(self, zone_change, **kwargs):
 
         return response == 200
 
+    def wait_until_zone_synced(self, zone_change):
```
Comment (mitruly) [inline modules/api/functional_test/vinyldns_python.py]: Can we consolidate this with `wait_until_zone_change_status_synced`?

File: modules/api/functional_test/live_tests/zones/create_zone_test.py [hunk 9]
```diff
@@ -63,7 +63,8 @@ def test_create_zone_success(shared_zone_test_context):
         }
         result = client.create_zone(zone, status=202)
         result_zone = result['zone']
-        client.wait_until_zone_change_status(result, 'Synced')
+        client.wait_until_zone_change_status_synced(result)
```
Comment (rebstar6) [inline modules/api/functional_test/live_tests/zones/create_zone_test.py]: this is still super odd. were you able to get to the bottom of why the change is appearing completed before the zone has finished syncing?

Comment (rebstar6) [reply modules/api/functional_test/live_tests/zones/create_zone_test.py]: ahh i see it. look here https://github.com/vinyldns/vinyldns/blob/5a02347cfb5a8bd92b503586d7ca992a7dabf1f1/modules/api/src/main/scala/vinyldns/api/backend/CommandHandler.scala#L139

That 1st call to zoneChangeProcessor(sync) not only saves the zone in syncing status, but also saves the change as synced even though it's not complete. hrmm. should probably dig more, most likely that initial save should save the zone status, but store the change as pending still.

Comment (rebstar6) [reply modules/api/functional_test/live_tests/zones/create_zone_test.py]: ^^ said, i think what were doing to address that in this PR is reasonable. i do wonder if it's causing any other transient failures though. can you make a card to investigate further?

File: modules/api/functional_test/vinyldns_python.py [hunk 10]
```diff
@@ -764,6 +724,23 @@ def wait_until_zone_exists(self, zone_change, **kwargs):
...
+        """
+        Waits a period of time for the zone sync to complete.
+
+        :param zone_change: the create zone change for the zone that has been created.
+        """
+        zone_id = zone_change[u'zone'][u'id']
+        retries = MAX_RETRIES
+        something = self.get_zone(zone_id)
```
Comment (rebstar6) [inline modules/api/functional_test/vinyldns_python.py]: probably wanna call this something other than `something` :P

File: modules/api/functional_test/live_tests/zones/create_zone_test.py (hunk 9 above)
Comment (britneywright) [reply modules/api/functional_test/live_tests/zones/create_zone_test.py]: Correct me if I'm misunderstanding, but in the `processChangeRequests` https://github.com/vinyldns/vinyldns/blob/master/modules/api/src/main/scala/vinyldns/api/backend/CommandHandler.scala#L139 we change the zone_change status to Synced, then actually sync the zone in the next step.

Comment (rebstar6) [reply modules/api/functional_test/live_tests/zones/create_zone_test.py]: thats what im seeing too - we save the change as synced 2x. on the zone status side, i believe the 1st save is saving with `zone status=Syncing` and the final one is `zone status=Active`. so a wait for zone status=Active (which should occur at the same time the latestSync happens) should work.

all this said, i think remove the wait for change here, and just update `wait_until_zone_synced` to be called `wait_until_zone_active` which waits for the full create (create + sync) to occur.

**Comment by britneywright** [comment]
`test_zone_sync_success` is failing here https://github.com/vinyldns/vinyldns/blob/master/modules/api/functional_test/live_tests/zones/sync_zone_test.py#L157 instead of returning a 403 on the zone sync it returns a 202. My thought is the previous sync is happening fast enough that the next sync is allowed to go through as well.

File: modules/api/functional_test/vinyldns_python.py (hunk 5 above)
Comment (nimaeskandary) [reply modules/api/functional_test/vinyldns_python.py]: Potential for race conditions where the latest change in the db is already synced and the last request is slow to be processed?

Comment (nimaeskandary) [reply modules/api/functional_test/vinyldns_python.py]: nevermind didn't notice its going after a specific zone change

Comment (nimaeskandary) [reply modules/api/functional_test/vinyldns_python.py]: Plusone-ing rebeccas comment on exiting loop early if status is failed, `while (latest_change[u'status'] != 'Synced' and retries > 0) and latest_change[u'status'] != 'Failed':`

**Comment by nimaeskandary** [comment]
thumbs except rebeccas comment regarding early exit on sync status

File: modules/api/functional_test/vinyldns_python.py (hunk 5 above)
Comment (britneywright) [reply modules/api/functional_test/vinyldns_python.py]: I forgot to comment or maybe I put it in an outdated place, for some reason when I added the latest_change[u'status'] != 'Failed' the func tests took 5 minutes longer. That happened multiple times so removed it.

Comment (britneywright) [reply modules/api/functional_test/vinyldns_python.py]: ah yeah that was what this comment was in reference too https://github.com/vinyldns/vinyldns/pull/373#issuecomment-444347708

Comment (nimaeskandary) [reply modules/api/functional_test/vinyldns_python.py]: @britneywright with an `OR` I could see it adding time depending on how its combined with retries, have you tried with an `AND`?

Comment (britneywright) [reply modules/api/functional_test/vinyldns_python.py]: We have no tests that are set up to have a Failed zone change status so we don't find ourselves getting a Failed and don't need to exit early. If we're still looping it's because the status is still Pending and not Synced yet.

Comment (britneywright) [reply modules/api/functional_test/vinyldns_python.py]: I added it, same run time, all the tests still pass.

File: modules/api/functional_test/live_tests/zones/sync_zone_test.py [hunk 11]
```diff
@@ -151,10 +150,7 @@ def test_sync_zone_success(shared_zone_test_context):
 
         # sync again
         change = client.sync_zone(zone['id'], status=202)
-        client.wait_until_zone_change_status(change, 'Synced')
-
-        # confirm cannot again sync without waiting
-        client.sync_zone(zone['id'], status=403)
+        client.wait_until_zone_change_status_synced(change)
```
Comment (rebstar6) [inline modules/api/functional_test/live_tests/zones/sync_zone_test.py]: even this technically isnt waiting enough logically. could be moving on too early

File: modules/api/functional_test/vinyldns_python.py [hunk 12]
```diff
@@ -764,6 +726,23 @@ def wait_until_zone_exists(self, zone_change, **kwargs):
...
+        """
+        Waits a period of time for the zone sync to complete.
+
+        :param zone_change: the create zone change for the zone that has been created.
+        """
+        zone_id = zone_change[u'zone'][u'id']
+        retries = MAX_RETRIES
+        zone = self.get_zone(zone_id)
+
+        while u'latestSync' not in zone and retries > 0:
```
[the remark this answers is not in the document]
Comment (britneywright) [reply modules/api/functional_test/vinyldns_python.py]: change is here https://github.com/vinyldns/vinyldns/pull/373/files#diff-db0cbd6424228193b3a89af595a286a8R744

[the remark this answers is not in the document]
Comment (britneywright) [reply modules/api/functional_test/vinyldns_python.py]: ah I see what you're saying, get the zone in the test rather than in this function.

File: modules/api/functional_test/live_tests/zones/sync_zone_test.py (hunk 11 above)
Comment (rebstar6) [reply modules/api/functional_test/live_tests/zones/sync_zone_test.py]: remove the wait_until_zone_change_status_synced, wait_for_zone_sync_to_complete will cover it

File: modules/api/functional_test/vinyldns_python.py [hunk 13]
```diff
@@ -2,6 +2,7 @@
 import time
 import logging
 import collections
+import datetime
```
Comment (rebstar6) [inline modules/api/functional_test/vinyldns_python.py]: dont think this is used

File: modules/api/functional_test/vinyldns_python.py [hunk 14]
```diff
@@ -700,30 +667,23 @@ def wait_until_recordset_deleted(self, zone_id, record_set_id, **kwargs):
...
+        latest_change = zone_change
         retries = MAX_RETRIES
-        while change[u'status'] != expected_status and retries > 0:
-            history = self.get_zone_history(zone_change[u'zone'][u'id'])
-
-            if u'zoneChanges' in history:
-                zone_changes = history[u'zoneChanges']
-                matching_changes = filter(change_id_match, zone_changes)
 
+        while (latest_change[u'status'] != 'Synced' and retries > 0) and latest_change[u'status'] != 'Failed':
```
Comment (rebstar6) [inline modules/api/functional_test/vinyldns_python.py]: parens shouldnt matter for `and`

File: modules/api/functional_test/vinyldns_python.py [hunk 15]
```diff
@@ -736,7 +696,9 @@ def wait_until_zone_deleted(self, zone_id, **kwargs):
         retries = MAX_RETRIES
         url = urljoin(self.index_url, u'/zones/{0}'.format(zone_id))
         response, data = self.make_request(url, u'GET', self.headers, not_found_ok=True, status=(200, 404), **kwargs)
+        print("RESPONSE: " + str(response))
         while response != 404 and retries > 0:
+            print("RESPONSE: " + str(response))
```
Comment (rebstar6) [inline modules/api/functional_test/vinyldns_python.py]: do we need these prints? ^^

File: modules/api/functional_test/vinyldns_python.py [hunk 16]
```diff
@@ -764,6 +726,22 @@ def wait_until_zone_exists(self, zone_change, **kwargs):
...
+    def wait_until_zone_active(self, zone_id):
+        """
+        Waits a period of time for the zone sync to complete.
+
+        :param zone_id: the ID for the zone.
+        """
+        retries = MAX_RETRIES
+        zone_request = self.get_zone(zone_id)
+
+        while ('status' not in zone_request or zone_request[u'zone'][u'status'] != 'Active') and retries > 0:
```
Comment (rebstar6) [inline modules/api/functional_test/vinyldns_python.py]: should this be
`'zone' not in zone_request`
or
`'status' not in zone_request[u'zone']`
?
